File: packages/php/src/TokenRecord.php
<?php

declare(strict_types=1);

namespace NebulaToken;

/**
 * Server-side record — one row per issued token ([N-10]).
 *
 * Immutable: a store hands out snapshots and applies its own writes, so no
 * caller can mutate a row it was merely shown. The lifecycle transitions are
 * the two `with*` helpers below.
 */
final readonly class TokenRecord
{
    /**
     * @param string      $verifierHash Lowercase hex HMAC — never the raw verifier ([N-14]).
     * @param string|null $deviceIdHash Lowercase hex HMAC, or null when unbound.
     * @param int         $createdAt    Unix seconds ([N-2]).
     */
    public function __construct(
        public string $selector,
        public string $verifierHash,
        public string $kid,
        public string $familyId,
        public int $generation,
        public string $userId,
        public ?string $deviceIdHash,
        public int $createdAt,
        public int $familyExpiresAt,
        public int $idleExpiresAt,
        public TokenStatus $status = TokenStatus::Active,
        public ?int $rotatedAt = null,
        public ?string $replacedBySelector = null,
    ) {
    }

    /** The `rotated` transition. On a grace retry `$rotatedAt` keeps its original value ([N-10]). */
    public function asRotated(int $rotatedAt, string $replacedBySelector): self
    {
        return new self(
            $this->selector, $this->verifierHash, $this->kid, $this->familyId,
            $this->generation, $this->userId, $this->deviceIdHash, $this->createdAt,
            $this->familyExpiresAt, $this->idleExpiresAt, TokenStatus::Rotated,
            $rotatedAt, $replacedBySelector,
        );
    }

    /** The `revoked` transition: status only — `replacedBySelector` is retained ([N-15]). */
    public function asRevoked(): self
    {
        return new self(
            $this->selector, $this->verifierHash, $this->kid, $this->familyId,
            $this->generation, $this->userId, $this->deviceIdHash, $this->createdAt,
            $this->familyExpiresAt, $this->idleExpiresAt, TokenStatus::Revoked,
            $this->rotatedAt, $this->replacedBySelector,
        );
    }
}

File: packages/java/src/main/java/dev/nebulatoken/TokenRecord.java
package dev.nebulatoken;

/**
 * The server-side record -- one row per issued token ([N-10]).
 *
 * <p>Immutable, so a store may hand the same instance to several threads and
 * the engine can never mutate what the store holds. The two state transitions a
 * store needs are {@link #rotated} and {@link #revoked}, which return new
 * instances.
 *
 * <p>Nothing here is a raw secret: {@code verifierHash} and {@code deviceIdHash}
 * are keyed digests, so the generated {@code toString} is safe to log ([N-14]).
 * The token itself and the raw device identifier never reach this type.
 *
 * @param selector           primary key; the only token-derived value that may be indexed ([N-45])
 * @param verifierHash       lowercase hex HMAC-SHA-256(pepper[kid], verifier bytes)
 * @param kid                pepper id used for BOTH hashes on this row
 * @param familyId           lowercase hex of 16 CSPRNG bytes, fixed at login
 * @param generation         0 at issue, +1 per rotation
 * @param userId             owner
 * @param deviceIdHash       lowercase hex HMAC-SHA-256(pepper[kid], "device:" + deviceId), or null when unbound
 * @param createdAt          issue time of this record, unix seconds ([N-2])
 * @param familyExpiresAt    absolute deadline, fixed at login; MUST never be extended
 * @param idleExpiresAt      sliding deadline: min(now + idleTtl, familyExpiresAt)
 * @param status             active / rotated / revoked
 * @param rotatedAt          set on first rotation; a grace retry MUST keep the original value ([N-30])
 * @param replacedBySelector selector of the successor record
 */
public record TokenRecord(
        String selector,
        String verifierHash,
        String kid,
        String familyId,
        int generation,
        String userId,
        String deviceIdHash,
        long createdAt,
        long familyExpiresAt,
        long idleExpiresAt,
        TokenStatus status,
        Long rotatedAt,
        String replacedBySelector) {

    /** This record, rotated. Used by stores implementing the [N-17] compare-and-set. */
    public TokenRecord rotated(long newRotatedAt, String successorSelector) {
        return new TokenRecord(selector, verifierHash, kid, familyId, generation, userId,
                deviceIdHash, createdAt, familyExpiresAt, idleExpiresAt,
                TokenStatus.ROTATED, newRotatedAt, successorSelector);
    }

    /**
     * This record, revoked. {@code rotatedAt} and {@code replacedBySelector} are
     * carried over untouched -- revocation must not erase the rotation lineage
     * that reuse detection reads ([N-15]).
     */
    public TokenRecord revoked() {
        return new TokenRecord(selector, verifierHash, kid, familyId, generation, userId,
                deviceIdHash, createdAt, familyExpiresAt, idleExpiresAt,
                TokenStatus.REVOKED, rotatedAt, replacedBySelector);
    }
}
